File: DKImagePickerControllerDemo_OC/DKImagePickerControllerDemo_OC/Source/DKImageManger/DKBaseManager.m
//
//  DKBaseManager.m
//  DKImagePickerControllerDemo_OC
//
//  Created by zm on 2017/7/6.
//  Copyright © 2017年 zm. All rights reserved.
//

#import "DKBaseManager.h"

@interface DKBaseManager()

@end
@implementation DKBaseManager
- (instancetype)init{
    if (self = [super init]) {
        _observers = [NSHashTable weakObjectsHashTable];
    }
    return self;
    
}


- (void)addObserver:(id)object{
    [self.observers addObject:object];
}

- (void)removeObserver:(id)object{
    [self.observers removeObject:object];
}

- (void)notifyObserversWithSelector:(SEL)sel
                             object:(id)object{
    [self notifyObserversWithSelector:sel object:object objectTwo:nil];
    
}



#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
- (void)notifyObserversWithSelector:(SEL)selector
                             object:(id)object
                          objectTwo:(id)objectTwo{
    if (self.observers.count > 0) {
        dispatch_async(dispatch_get_main_queue(), ^{
            for (id anyObject in self.observers.allObjects) {
                if ([anyObject respondsToSelector:selector]) {
                   [anyObject performSelector:selector withObject:object withObject:objectTwo];
                }
            }
        });
    }
}


@end

File: DKImagePickerController/DKBaseManager.swift
//
//  DKBaseManager.swift
//  DKImagePickerController
//
//  Created by ZhangAo on 17/11/2017.
//  Copyright © 2017 ZhangAo. All rights reserved.
//

import Foundation

open class DKBaseManager: NSObject {
    
    private let observers = NSHashTable<AnyObject>.weakObjects()
    
    open func add(observer object: AnyObject) {
        self.observers.add(object)
    }
    
    open func remove(observer object: AnyObject) {
        self.observers.remove(object)
    }
    
    open func notify(with selector: Selector, object: AnyObject?) {
        self.notify(with: selector, object: object, objectTwo: nil)
    }
    
    open func notify(with selector: Selector, object: AnyObject?, objectTwo: AnyObject?) {
        if self.observers.count > 0 {
            DispatchQueue.main.async(execute: {
                for observer in self.observers.allObjects {
                    if observer.responds(to: selector) {
                        _ = observer.perform(selector, with: object, with: objectTwo)
                    }
                }
            })
        }
    }
    
}
